README.md

```diff
@@ -2,253 +2,220 @@
--t: Just test you configuration
-```
+For a minimal stack with PostgreSQL you need the following environment variables to setup the connection to PostgreSQL and initializing the database:
 
-If you want to enforce an update, create a `/opt/opennms/etc/do-upgrade` file.
-Starting with `-i` or `-s` will run the `install -dis` command once to update the configuration and database schema.
+```bash
+# Environment variables for OpenNMS database connection
+TZ=Europe/Berlin
+OPENNMS_DATABASE_NAME=opennms
+OPENNMS_DATABASE_USER=opennms
+OPENNMS_DATABASE_PASSWORD=opennms
+POSTGRES_HOST=database
+POSTGRES_PORT=5432
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=postgres
+```
 
-All options which do upgrades or start OpenNMS Horizon verify if the configuration is valid and pass the configuration test.
+Ensure the time zone `TZ` is set to the correct time zone you're in.
+The examples use `.opennms.env` and `.postgres.env` files which contain these settings.
+You can also set them in the docker-compose file if you want.
 
-## Using etc-overlay for custom configuration
+### Persistence your data
 
-If you just want to maintain custom configuration files outside of OpenNMS, you can use an etc-overlay directory.
-All files in this directory are just copied into /opt/opennms/etc in the running container.
-You can just mount a local directory like this:
+Running the container images are ephemeral and by default do not persist data.
+The following directories are required to be persisted in a volume or as a bind-mount depending on your environment.
+The minimal example uses volumes for data which is not required to be edited on the file system and bind-mounts for configurations.
 
-```yml
-volumes:
-  - ./etc-overlay:/opt/opennms-etc-overlay
-```
+* PostgreSQL database: `- data-pgsql:/var/lib/postgresql/data`
+* OpenNMS Horizon JasperReports: `- data-opennms-reports:/opt/opennms/share/reports`
+* OpenNMS Horizon RRD files: `- data-opennms-rrd:/opt/opennms/share/rrd`
+* OpenNMS Horizon MIB files: `- data-opennms-mibs:/opt/opennms/share/mibs`
+* OpenNMS configuration files: `- ./opennms-etc:/opt/opennms/etc`
+      
 
-## Using jetty-overlay to customize Jetty WEBINF
+For the reason some configuration files from OpenNMS Horizon need to be edited and migrated to newer versions from the file system it is recommended you persist the `opennms/etc` directory on your local files system so you can easily access it or set it under version control.
 
-If you just want to maintain custom configuration files for the Jetty application container, you can use an jetty-overlay directory.
-All files in this directory are just copied into /opt/opennms/jetty-webapps/opennms/WEB-INF in the running container.
-You can just mount a local directory like this:
+### Configuration through Environment Variables
 
-```yml
-volumes:
-  - ./jetty-overlay:/opt/opennms-jetty-webinf-overlay
-```
+OpenNMS Horizon has no built-in mechanism to get configured through Environment variables.
```

Comment:
Literally, it is still the case, OpenNMS itself is not container aware and not configurable through environment variables, which is the reason we use confd to bridge the gap between the container world and the legacy OpenNMS config world.

Just as a side note, we need to be careful to stay away from "runtime" configuration files like snmp-config.xml or anything else which can be changed from the Web UI at runtime. Another thing is Karaf itself, which is by design not built to run in containers (immutable processes), but this is another topic.

Reply:
> My opinion is that key names should resemble the property name they are associated with as closely as possible. This makes it easy to "reverse engineer" a key name for a given property both when troubleshooting as well as when adding new keys.

I've tried to find a compromise between writeable environment variables without having too much redundancy with writing always org.opennms ... Consistency and naming is not a strength in a 20-year-old project :) But as it is just about names, I'll follow what fits better for you. As I told already @agalue, I'm from Germany and I'm used to long names :)

Reply:
If we decide we like confd and want to commit further to it in the future we should be able to template runtime-configurable files as well. The process that is modifying the XML would need to be updated to instead write new key-values to the keystore.

In this case we would either run confd in daemon mode so it will see the key changes and automatically refresh the updated file, or we would trigger a `onetime` refresh of the config file to reflect the new key values.

Ultimately if we go down this road we do away with all config file editing and make all config changes go through the keystore. This means no more manual modifications of XML files (which I heard is a pain point right now). In the past I've done exactly this, but it requires a significant commitment and confidence in the strategy as it is a pretty big change. Not saying we are ready for this or want it right now, but I think we should think about it.

Here is a couple benefits I found when working with confd in the past that I think would be applicable to us:

Helps "cloudify" the app. For instance if we want to run our application in the cloud often times you will be running on a VM or container that is ephemeral and does not necessarily have any persistent storage. In cases like this you run your keystore separately somewhere where it does have persistent storage. This way you are free to blow away your application container at any time and it will always come back with the exact same configuration. You can also snapshot your keystore at any given time and have the entire config for that install for that time to restore from (nice for support to be able to restore entire config in one step :)).

Helps initial configuration or creating wizards for config. On a previous project I worked on, the application was instantiated in a cloud using a "VNFM" (virtual network function manager). At instantiation time you could configure the application right from the VNFM web gui before spinning it up since configuration was just setting key-value pairs in a keystore. You can even write simple wizards for common configuration use cases. Operators in this case never have to edit XML manually so it reduces the human error element significantly and makes configuration more intuitive (in my opinion).

Reply:
> If we decide we like confd and want to commit further to it in the future we should be able to template runtime-configurable files as well. The process that is modifying the XML would need to be updated to instead write new key-values to the keystore.
> 
> In this case we would either run confd in daemon mode so it will see the key changes and automatically refresh the updated file, or we would trigger a `onetime` refresh of the config file to reflect the new key values.
> 
> Ultimately if we go down this road we do away with all config file editing and make all config changes go through the keystore. This means no more manual modifications of XML files (which I heard is a pain point right now). In the past I've done exactly this, but it requires a significant commitment and confidence in the strategy as it is a pretty big change. Not saying we are ready for this or want it right now, but I think we should think about it.

The main design issue here is, we persist configuration files persisted on the file system which needs to be injected into the image, which can be modified from the Web application at runtime and does not get synched back to your confd configuration source. This pattern just doesn't work this type of things we need a different approach here. Having confd running in background doesn't help here. If you set something in the WebUI it is not persisted in your configuration backend used by confd.

To address this issue runtime configurations would be better stored in a network storage, so the application can run as immutable process and we don't have a dependency on the nodes local file system anymore.

For startup configuration like our properties, this is not such a big deal, cause they have to be set on startup and to change them you have to restart the process and there is no other way to change them.

Reply:
> Helps initial configuration or creating wizards for config. On a previous project I worked on, the application was instantiated in a cloud using a "VNFM" (virtual network function manager). At instantiation time you could configure the application right from the VNFM web gui before spinning it up since configuration was just setting key-value pairs in a keystore. You can even write simple wizards for common configuration use cases. Operators in this case never have to edit XML manually so it reduces the human error element significantly and makes configuration more intuitive (in my opinion).

Yes totally agree with you here, I think it should be strategical decision refactoring our product to make it work exactly like this.

Reply:
> To address this issue runtime configurations would be better stored in a network reachable versioned storage, so the application can run as immutable process and we don't have a dependency on the nodes local file system anymore.

Moving those runtime-configurable files to networked storage is one solution. Another could be to change our application so that it no longer writes directly to any configuration files, it only writes to its keystore. This could be significant work depending on how much of that there is going on (guessing a lot). Ideally all configuration is derived from one location, the keystore, which is just a bunch of key-values in a hierarchy.

Note that dumping an entire XML or JSON blob as a key's value is an option (seen this before), although it is not ideal. Can also break it up a bit so that some elements are associated with some keys rather than a whole XML file for instance. This can be used as a bridging step if the intent was to transition to all config derived from keys.